src/Microsoft.AspNetCore.Routing/DefaultLinkGenerator.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Encodings.Web;
+using Microsoft.AspNetCore.Routing.Internal;
+using Microsoft.AspNetCore.Routing.Matchers;
+using Microsoft.AspNetCore.Routing.Template;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.ObjectPool;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    public class DefaultLinkGenerator : ILinkGenerator
+    {
+        private readonly IEndpointFinder _endpointFinder;
+        private readonly ObjectPool<UriBuildingContext> _uriBuildingContextPool;
+        private readonly ILogger<DefaultLinkGenerator> _logger;
+
+        public DefaultLinkGenerator(
+            IEndpointFinder endpointFinder,
+            ObjectPool<UriBuildingContext> uriBuildingContextPool,
+            ILogger<DefaultLinkGenerator> logger)
+        {
+            _endpointFinder = endpointFinder;
+            _uriBuildingContextPool = uriBuildingContextPool;
+            _logger = logger;
+        }
+
+        public string GetLink(LinkGeneratorContext context)
+        {
+            if (TryGetLink(context, out var link))
+            {
+                return link;
+            }
+
+            throw new InvalidOperationException("Could not find a matching endpoint to generate a link.");
+        }
+
+        public bool TryGetLink(LinkGeneratorContext context, out string link)
+        {
+            var address = context.Address;
+            var endpoints = _endpointFinder.FindEndpoints(address);
+            link = null;
+
+            if (endpoints == null)
+            {
+                return false;
+            }
+
+            var matcherEndpoints = endpoints.OfType<MatcherEndpoint>();
+            if (!matcherEndpoints.Any())
+            {
+                //todo:log here
+                return false;
+            }
+
+            foreach (var endpoint in matcherEndpoints)
+            {
+                link = GetLink(endpoint.RouteTemplate, endpoint.Values, context);
+                if (link != null)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private string GetLink(
+            RouteTemplate template,
+            IReadOnlyDictionary<string, object> defaultValues,
+            LinkGeneratorContext context)
+        {
+            var defaults = new RouteValueDictionary(defaultValues);
+            var templateBinder = new TemplateBinder(
```

Comment:
Ideally this can be constructed once per matcherendpoint, but wanted to send out this PR to see if the general direction is correct.